File: sdk/e2e/02_update_router_invalid_test.py
import os
import logging

import requests
import turing

from turing.router.config.experiment_config import ExperimentConfig
from turing.router.config.resource_request import ResourceRequest
from turing.router.config.enricher import Enricher
from turing.router.config.common.env_var import EnvVar
from turing.router.config.router_version import RouterStatus


def test_update_router_invalid_config():
    # get the existing router that has been created in 01_create_router_test.py
    logging.info("Retrieving router...")
    router = turing.Router.get(1)
    assert router is not None

    # get the router config from the deployed router
    new_router_config = router.config

    # set up the new experiment config
    new_router_config.experiment_config = ExperimentConfig(
        type="nop",
    )

    # set up the new senricher for the router
    new_router_config.enricher = Enricher(
        image=os.getenv("TEST_ECHO_IMAGE"),
        resource_request=ResourceRequest(
            min_replica=1, max_replica=1, cpu_request="10", memory_request="1Gi"
        ),
        endpoint="anything",
        timeout="2s",
        port=80,
        env=[EnvVar(name="TEST_ENV", value="enricher")],
    )

    # update router
    logging.info("Updating router with new config...")
    router.update(new_router_config)

    # get router version 2
    logging.info("Waiting for new router version to be deployed/fail...")
    try:
        router.wait_for_version_status(RouterStatus.FAILED, 2)
    except TimeoutError:
        raise Exception(
            f"Turing API is taking too long for router {router.id} with version 2 to get deployed."
        )
    router_ver_2 = router.get_version(2)
    assert router_ver_2.status == RouterStatus.FAILED

    # ensure router does not get updated to the failed version (version 2)
    logging.info(
        "Ensuring the router does not get updated to the failed version (i.e. the version number remains as 1)..."
    )
    router = turing.Router.get(1)
    assert router.version == 1

    # test router endpoint by posting a single request
    assert (
        router.endpoint
        == f'http://{router.name}-turing-router.{os.getenv("PROJECT_NAME")}.{os.getenv("KSERVICE_DOMAIN")}/v1/predict'
    )
    logging.info("Testing router endpoint...")
    response = requests.post(
        url=router.endpoint,
        headers={
            "Content-Type": "application/json",
            "X-Mirror-Body": "true",
        },
        json={"client": {"id": 4}},
    )
    assert response.status_code == 200
    expected_response = {
        "experiment": {
            "configuration": {"foo": "foo", "route_name": "control"},
        },
        "route_responses": [
            {"data": {"version": "control"}, "is_default": False, "route": "control"}
        ],
    }
    assert response.json()["response"] == expected_response

File: api/e2e/test/02_update_router_invalid_test.go
//go:build e2e

package e2e

import (
	"bytes"
	"fmt"
	"net/http"
	"path/filepath"
	"testing"

	"github.com/caraml-dev/turing/api/turing/models"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"github.com/tidwall/gjson"
)

/*
Pre:
testCreateRouter run successfully.
Steps:
Update existing router with invalid config - High resource requests
a. Wait for failed deployment
b. GET router version 2 > shows status "failed"
c. GET router > config still shows version 1
d. Test cluster state that all deployments exist
e. Make a request to the router, validate the response.
*/
func TestUpdateRouterInvalidConfig(t *testing.T) {
	// Read existing router that MUST have already exists from previous create router e2e test
	// Router name is assumed to follow this format: e2e-experiment-{{.TestID}}
	routerName := "e2e-experiment-" + globalTestContext.TestID
	t.Logf("Retrieving router with name '%s' created from previous test step", routerName)
	existingRouter, err := getRouterByName(
		globalTestContext.httpClient, globalTestContext.APIBasePath, globalTestContext.ProjectID, routerName)
	require.NoError(t, err)

	// Read router config test data
	data := makeRouterPayload(filepath.Join("testdata", "update_router_high_cpu.json.tmpl"), globalTestContext)

	// Update router
	url := fmt.Sprintf(
		"%s/projects/%d/routers/%d",
		globalTestContext.APIBasePath,
		globalTestContext.ProjectID,
		existingRouter.ID,
	)
	t.Log("Updating router: PUT " + url)
	req, err := http.NewRequest("PUT", url, bytes.NewReader(data))
	require.NoError(t, err)
	req.Header.Set("Content-Type", "application/json")
	resp, err := globalTestContext.httpClient.Do(req)
	require.NoError(t, err)
	assert.Equal(t, http.StatusOK, resp.StatusCode, readBody(t, resp))

	// Until deployment timeout, wait for router version status to change
	t.Log("Waiting for router version to deploy")
	err = waitDeployVersion(
		globalTestContext.httpClient,
		globalTestContext.APIBasePath,
		globalTestContext.ProjectID,
		int(existingRouter.ID),
		2,
	)
	assert.NoError(t, err)

	// Get router version 2
	t.Log("Testing GET router version")
	routerVersion, err := getRouterVersion(
		globalTestContext.httpClient,
		globalTestContext.APIBasePath,
		globalTestContext.ProjectID,
		int(existingRouter.ID),
		2,
	)
	require.NoError(t, err)
	assert.Equal(t, models.RouterVersionStatusFailed, routerVersion.Status)
	assert.NotEmpty(t, routerVersion.Error)

	t.Log("Ensure existing router do not update the version to failed version i.e the version still unchanged at 1")
	router, err := getRouter(
		globalTestContext.httpClient,
		globalTestContext.APIBasePath,
		globalTestContext.ProjectID,
		int(existingRouter.ID),
	)
	require.NoError(t, err)
	require.NotNil(t, router.CurrRouterVersion)
	assert.Equal(t, 1, int(router.CurrRouterVersion.Version))

	downstream, err := getRouterDownstream(globalTestContext.clusterClients,
		globalTestContext.ProjectName,
		fmt.Sprintf("%s-turing-router", router.Name))
	assert.NoError(t, err)
	assert.Equal(t, downstream, fmt.Sprintf("%s-turing-router-%d.%s.%s",
		router.Name, 1, globalTestContext.ProjectName, globalTestContext.KServiceDomain))

	// Check that previous enricher, router, ensembler deployments exist
	t.Log("Checking cluster state")
	assert.True(t, isDeploymentExists(
		globalTestContext.clusterClients,
		globalTestContext.ProjectName,
		fmt.Sprintf("%s-turing-router-%d-0-deployment", router.Name, 1)))
	assert.True(t, isDeploymentExists(
		globalTestContext.clusterClients,
		globalTestContext.ProjectName,
		fmt.Sprintf("%s-turing-enricher-%d-0-deployment", router.Name, 1)))
	assert.True(t, isDeploymentExists(
		globalTestContext.clusterClients,
		globalTestContext.ProjectName,
		fmt.Sprintf("%s-turing-ensembler-%d-0-deployment", router.Name, 1)))

	// Make request to router
	t.Log("Testing router endpoint")
	router, err = getRouter(
		globalTestContext.httpClient,
		globalTestContext.APIBasePath,
		globalTestContext.ProjectID,
		int(router.ID),
	)
	require.NoError(t, err)
	assert.Equal(t,
		fmt.Sprintf(
			"http://%s-turing-router.%s.%s/v1/predict",
			router.Name,
			globalTestContext.ProjectName,
			globalTestContext.KServiceDomain,
		),
		router.Endpoint,
	)

	withRouterResponse(t,
		http.MethodPost,
		router.Endpoint,
		http.Header{
			"Content-Type":  {"application/json"},
			"X-Mirror-Body": {"true"},
		},
		`{"client": {"id": 4}}`,
		func(response *http.Response, responsePayload []byte) {
			assert.Equal(t, http.StatusOK, response.StatusCode,
				"Unexpected response (code %d): %s",
				response.StatusCode, string(responsePayload))
			actualResponse := gjson.GetBytes(responsePayload, "response").String()
			expectedResponse := `{
			  "experiment": {
				"configuration": {
					"foo":"bar",
					"route_name":"treatment-a"
				}
			  },
			  "route_responses": [
				{
				  "data": {
					"version": "control"
				  },
				  "is_default": false,
				  "route": "control"
				}
			  ]
			}`
			assert.JSONEq(t, expectedResponse, actualResponse)
		})
}
